### PR TITLE
Fix allocation of LVGL display buffer memory when <60kB memory remaining (AEGHB-5)

### PR DESCRIPTION
**Pull request by dvosully**
The LVGL initialisation code fails catastrophically when the available internal memory is below 60kB. When allocating the display buffer memory, it tries to leave 60kB free, doesn't handle the wraparaound from 55kB-60kB or similar, and tries to allocate way more memory than ever exists.

This PR changes the behaviour to check whether there is sufficient memory available to leave (60kB - amount now configurable) available, and if not, splits the different and allocates half of the remaining memory to the display buffer.

It also adds the option to configure the amount of stack allocated to the LVGL task (the default of 8kB is overkill in many applications), the priority of the LVGL task, and the core affinity of the LVGL task. The default values remain the same.

### CODE REVIEW

File: components/gui/lvgl_gui/lvgl_gui.c [hunk 1]
```diff
@@ -79,9 +79,19 @@ static esp_err_t lvgl_display_init(scr_driver_t *driver)
...
+            // If we can't leave 60k of spare memory, just leave half of whatever is left.
+            remain_size = free_size / 2;
+            ESP_LOGW(TAG, "Final remain size = %d Bytes", remain_size);
+        }
+        if (alloc_pixel > DISP_BUF_SIZE) {
+            // If half the remaining memory is more than we originally planned to allocate, just stick with the original plan.
+            alloc_pixel = DISP_BUF_SIZE;
+        }
         size_t allow_size = (free_size - remain_size) & 0xfffffffc;
         alloc_pixel = SIZE_TO_PIXEL(allow_size / BUFFER_NUMBER);
```
Comment (Lzw655) [inline components/gui/lvgl_gui/lvgl_gui.c:96]:
```suggestion
        size_t allow_size = (free_size - remain_size) & 0xfffffffc;
        alloc_pixel = SIZE_TO_PIXEL(allow_size / BUFFER_NUMBER);
        if (alloc_pixel > DISP_BUF_SIZE) {
            // If half the remaining memory is more than we originally planned to allocate, just stick with the original plan.
            alloc_pixel = DISP_BUF_SIZE;
        }
```
I think here should be like this.

**Review by Lzw655 — CHANGES_REQUESTED** [review]
@dvosully Hi, thank you for fixing some problems in  `lvgl_gui`. I found a little place that needs to modify, PTAL. Others LGTM.

File: components/gui/lvgl_gui/lvgl_gui.c [hunk 2]
```diff
@@ -79,9 +79,19 @@ static esp_err_t lvgl_display_init(scr_driver_t *driver)
...
         ESP_LOGW(TAG, "Exceeded max free size, force shrink to %u Byte", allow_size);
```
Comment (Lzw655) [inline components/gui/lvgl_gui/lvgl_gui.c:97]: Please also update here.

Comment (dvosully) [reply components/gui/lvgl_gui/lvgl_gui.c]: You are correct, the order of the check was wrong. I have incorporated your suggested changes in the updated branch.

I have also moved the warning message to only display if the buffer size has been changed.

Thanks for the feedback.

**Review by Lzw655 — APPROVED** [review]

**Comment by Lzw655** [comment]
@dvosully Sorry for my delay. I was sick this week. And I'll merge it as soon.

**Comment by Lzw655** [comment]
Actually now we recommend using the [esp_lcd](https://github.com/espressif/esp-idf/tree/master/components/esp_lcd) driver instead of `dispaly` in iot-solution. And there are also some exist components like [esp_lvgl_port](https://components.espressif.com/components/espressif/esp_lvgl_port) and [esp_lcd_touch](https://components.espressif.com/components/espressif/esp_lcd_touch).

**Pull request closed without merging**